trillian/ctfe/dns_handlers.go

```diff
@@ -0,0 +1,324 @@
+)
+
+type dnsFunc func(context.Context, *CTDNSHandler, []string, dns.ResponseWriter, *dns.Msg)
+
+// CTDNSHandler encapsulates all the logic for serving CT related DNS
+// requests.
+type CTDNSHandler struct {
+	cfg      *configpb.LogConfig
+	logCtx   *LogContext
+	handlers []dnsHandler
+}
+
+type dnsHandler struct {
+	matchRE  *regexp.Regexp
+	handleFn dnsFunc
+}
+
+// NewDNS creates a new DNS handler for use with a particular CT log.
+func NewDNS(cfg *configpb.LogConfig, logCtx *LogContext) dns.Handler {
+	return &CTDNSHandler{
+		cfg:    cfg,
+		logCtx: logCtx,
+		handlers: []dnsHandler{
+			{matchRE: sthRE, handleFn: sthFunc},
+			{matchRE: consistRE, handleFn: consistFunc},
+			{matchRE: hashRE, handleFn: hashFunc},
+			{matchRE: treeRE, handleFn: treeFunc},
+		},
+	}
+}
+
+// ServeDNS implements the dns.Handler interface and is the main serving
+// API.
+func (c *CTDNSHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	// From the spec all requests must be class INET, type TXT.
+	if err := validate(r); err != nil {
+		glog.Warningf("CTDNSHandler.ServeDNS(): %v", err)
+		failWithRcode(w, r, dns.RcodeFormatError, err)
+		return
+	}
+
+	glog.V(1).Infof("CTDNSHandler.ServeDNS(): Query: %v", r.Question[0].Name)
+	q := dns.Fqdn(r.Question[0].Name)
+	ctx, cancel := context.WithTimeout(context.Background(), c.logCtx.instanceOpts.Deadline)
+	defer cancel()
+
+	// Find the first handler that accepts the name pattern match.
+	for _, h := range c.handlers {
+		params := h.matchRE.FindStringSubmatch(q)
+		// All our matches should include at least the zone as a capture group
+		// so we must have at least two results from the above.
+		if len(params) > 1 {
+			// Additionally check that the zone matched the last regex param and the
+			// whole string matched to avoid any false positives.
+			if params[0] == q && params[len(params)-1] == c.cfg.DnsZone {
+				// This handler accepted the match and provides the result.
+				h.handleFn(ctx, c, params, w, r)
+				return
+			}
+		}
+	}
+	// No handler matched. Reject the request.
+	failWithRcode(w, r, dns.RcodeNotZone, nil)
+}
+
+func sthFunc(ctx context.Context, c *CTDNSHandler, params []string, w dns.ResponseWriter, r *dns.Msg) {
+	sth, err := GetTreeHead(ctx, c.logCtx.rpcClient, c.logCtx.logID, c.logCtx.LogPrefix)
+	if err != nil {
+		failWithRcode(w, r, dns.RcodeServerFailure, err)
+		return
+	}
+
+	// Add the signature over the STH contents.
+	err = c.logCtx.signV1TreeHead(c.logCtx.signer, sth)
+	if err != nil || len(sth.TreeHeadSignature.Signature) == 0 {
+		failWithRcode(w, r, dns.RcodeServerFailure, err)
+		return
+	}
+
+	rr, err := buildSTHResponse(params[0], sth)
+	if err != nil {
+		failWithRcode(w, r, dns.RcodeServerFailure, err)
+		return
+	}
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Answer = append(m.Answer, rr)
+	if err := w.WriteMsg(m); err != nil {
+		glog.Warningf("sthFunc(): WriteMsg: %v", err)
+	}
+}
+
+// params will contain 0 = regex match text, 1 = start_index, 2 = first, 3 = second
```

Comment:
add consts for these indices to help with readability, or perhaps better return a struct?

Reply:
Same with the other parsing `.*Func` methods below

Reply:
Not sure about structs. I get a slice of strings back from the regex and just want them as integers. I can certainly improve the docs though.